ntoskrnl/cc/view.c

```diff
@@ -520,25 +519,27 @@ CcRosTrimCache(
             /* Reacquire the locks */
             oldIrql = KeAcquireQueuedSpinLock(LockQueueMasterLock);
             KeAcquireSpinLockAtDpcLevel(&current->SharedCacheMap->CacheMapLock);
+
+            /* Dereference the VACB */
+            CcRosVacbDecRefCount(current);
 #endif
         }
 
         /* Only keep iterating though the loop while the lock is held */
         current_entry = current_entry->Flink;
```

Comment:
Let's suppose that the code with the comment `/* Page out the VACB */` above could be run (I know that it's currently in the disabled part of the `#if 1`, so it currently is NOT run...).
We add a reference to the `current` VACB then release all the locks in order to run that "Page out the VACB" code.
I could imagine that during that time, some other code could reduce the `current` VACB ref count down to 1 (is it possible?). Then we reacquire all locks and remove our reference acquired on `current`. In that hypothetical scenario, its reference is now == 0.
`current_entry` is coming from a field (`VacbLruListEntry`) within that `current` VACB, whose reference count is now zero.
Because `CcRosVacbDecRefCount` destroys the VACB when the refcount reaches zero (see [here](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/cc/view.c;hb=90de2b660b12919c2f60f2f31ad41e37c123d662#l75) or [there](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/include/internal/cc.h;hb=90de2b660b12919c2f60f2f31ad41e37c123d662#l497)), when we get here doing `current_entry = current_entry->Flink;`, it seems to me that we would now reference invalid memory.
The original code was going around this problem because we had the `current` VACB still referenced while getting the `current_entry->Flink`.

Reply:
Note that in `CcRosFreeOneUnusedVacb` (your second commit), the `current_entry->Flink` dereferencing is done _while it was certain that VACB was in the list_, so I think that in that case it is OK to use it.

Reply:
The original code implies that freeing VACB on the `CcRosVacbDecRefCount` call shouldn't be possible here because freeing must be done outside of locks (see [a comment in CcRosFlushDirtyPages](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/cc/view.c;h=7c13afb3220f40a7d43f4ddbc79d2783a5c08483;hb=90de2b660b12919c2f60f2f31ad41e37c123d662#l405)).

Reply:
After looking at it again, I decided to move it up before the "Page out the VACB" code for consistency. Thanks for pointing out.